sycl/doc/OptionalDeviceFeatures.md

```diff
@@ -0,0 +1,719 @@
+a kernel may call, including functions that are called transitively from other
+functions.  In standard SYCL, device code is not allowed to contain function
+pointers, virtual functions, or indirect function calls.  It is therefore easy
+to compute the static call graph of a kernel.  By starting at the kernel
+function itself (e.g.  the function passed to `parallel_for`), the compiler can
+identify all functions called by that function, then it can find all functions
+called by those functions, etc.  Depending on the tool which does the analysis,
+the "static call graph" could include only those functions that reside in the
+same translation unit as the kernel, or it could include all functions that
+reside in the same executable image (or shared library) as the kernel.  In the
+sections below, we try to make the distinction clear whenever we refer to a
+kernel's static call graph.
+
+We are contemplating a DPC++ extension that would allow some limited use of
+function pointers in device code.  This feature is not yet fully defined or
+supported.  We expect that the semantics of this feature will include some way
+for the compiler to deduce a limited set of possible targets for each indirect
+function call.  Therefore, it is still possible for the compiler to construct a
+"static call graph" for each kernel, the only difference is that each call site
+now adds a set of possible target functions to a kernel's static call graph.
+The details about how this will work are expected to be included in the DPC++
+extension specification that enables indirect function calls.
+
+### An exported device function
+
+The term "exported device function" means a device function that is exported
+from a shared library as defined by [Device Code Dynamic Linking][5].
+
+[5]: <https://github.com/intel/llvm/blob/sycl/sycl/doc/SharedLibraries.md>
+
+### The FE compiler
+
+The term "FE compiler" refers to the entire DPC++ compiler chain that runs
+when the user executes the `clang++` command.  This includes the clang
+front-end itself, all passes over LLVM IR, the post-link tool, and any AOT
+compilation phases (when the user compiles in AOT mode).  The FE compiler does
+not include the JIT compiler which translates SPIR-V (or another IL format)
+into native code when the application executes.
+
+
+## Requirements
+
+There are several categories of requirements covered by this design.  Each of
+these is described in more detail in the sections that follow:
+
+* The FE compiler must issue a diagnostic in some cases when a kernel or device
+  function uses an optional feature.  However, the FE compiler must **not**
+  generate a diagnostic in other cases.
+
+* The runtime must raise an exception when a kernel using optional features
+  is submitted to a device that does not support those features.  This
+  exception must be raised synchronously from the kernel invocation command
+  (e.g. `parallel_for()`).
```

Comment:
This requirement is from the SYCL 2020 specification:

> Of course, applications that make use of optional kernel features should ensure that a kernel using such a feature is submitted only to a device that supports the feature. If the application submits a command group using a secondary queue, then any kernel submitted from the command group should use only features that are supported by both the primary queue’s device and the secondary queue’s device. If an application fails to do this, the implementation must throw a synchronous exception with the `errc::kernel_not_supported` error code from the kernel invocation command (e.g. `parallel_for()`).

https://www.khronos.org/registry/SYCL/specs/sycl-2020/html/sycl-2020.html#sec:optional-kernel-features

In general, our philosophy is to diagnose runtime errors by throwing an exception from the API that causes the error whenever this is reasonably possible.  We think this makes it easier for applications to know the cause of a failure.  For example, if the exception is thrown from the `submit()` call, it might not be clear if the error is from `paralle_for()` or from constructing an accessor, etc.

Is there a reason why you think the exception should be delayed and thrown from `submit()` rather than from `parallel_for()`?

Reply:
>> Is there a reason why you think the exception should be delayed and thrown from submit() rather than from parallel_for()?

It's related to how things are currently implemented. Currently we record everything that happens in the CG lambda(like "accessor created", "kernel execution requested") doing only basic API related checks, then when CG is executed we send this request to the scheduler component which may or may not construct low-level handles and enqueue low-level operations immediately(it depends on the situation). Most of the errors coming from the last step.

Also there are complains that `queue::submit` call is taking too much time of the calling thread, so eventually we might want to move as much as we can to a separate(dispatcher) thread. If we did so we would not be able to throw synchronous exceptions for most of things that currently happen during `queue::submit` anymore.

>> For example, if the exception is thrown from the submit() call, it might not be clear if the error is from paralle_for() or from constructing an accessor, etc.

CG lambda can have only one "action"(kernel execution or memcpy or ...), the exception "errc::kernel_not_supported" is quite specific, so I think it would be quite hard to not understand that it happens because `queue::submit` asks to execute an "incompatible" kernel.

Reply:
> CG lambda can have only one "action"(kernel execution or memcpy or ...), the exception "errc::kernel_not_supported" is quite specific, so I think it would be quite hard to not understand that it happens because queue::submit asks to execute an "incompatible" kernel.

I agree that it would be clear in this case.  However, there are other exceptions that could be thrown in the "submit" function, especially related to accessors.  The SYCL spec expects these to be thrown from the API that causes the error (e.g. from the `accessor` constructor if the parameters to the accessor are invalid), not delayed and thrown from the `queue::submit()` function.  These exceptions would not be clear if they were delayed because the submit function could construct many accessors.

Does DPC++ currently delay these exceptions also until the kernel submission function returns?

> Also there are complains that queue::submit call is taking too much time of the calling thread, so eventually we might want to move as much as we can to a separate(dispatcher) thread.

I'm curious about what we do now in the calling thread.  Does the JIT compilation happen currently in the calling thread, or is this already scheduled on a separate thread?  Note that we were careful to design the `errc::kernel_not_supported` exception such that it could be diagnosed in the calling thread even before attempting to JIT compile the kernel, thus allowing the JIT compilation to happen on a different thread.